packages/neuron-wallet/src/services/indexer/queue.ts

```diff
@@ -0,0 +1,221 @@
+  }
+
+  public indexLockHashes = async (lockHashes: string[]) => {
+    const lockHashIndexStates = await this.indexerRPC.getLockHashIndexStates()
+    const indexedLockHashes: string[] = lockHashIndexStates.map(state => state.lockHash)
+    const nonIndexedLockHashes = lockHashes.filter(i => !indexedLockHashes.includes(i))
+
+    await Utils.mapSeries(nonIndexedLockHashes, async (lockHash: string) => {
+      await this.indexerRPC.indexLockHash(lockHash)
+    })
+  }
+
+  // type: 'createdBy' | 'consumedBy'
+  public pipeline = async (lockHash: string, type: TxPointType, startBlockNumber: bigint) => {
+    let page = 0
+    let stopped = false
+    while (!stopped) {
+      const txs = await this.indexerRPC.getTransactionByLockHash(lockHash, page.toString(), this.per.toString())
+      if (txs.length < this.per) {
+        stopped = true
+      }
+      for (const tx of txs) {
+        let txPoint: CKBComponents.TransactionPoint | null = null
+        if (type === TxPointType.CreatedBy) {
+          txPoint = tx.createdBy
+        } else if (type === TxPointType.ConsumedBy) {
+          txPoint = tx.consumedBy
+        }
+
+        if (
+          txPoint &&
+          (BigInt(txPoint.blockNumber) >= startBlockNumber || this.tipBlockNumber - BigInt(txPoint.blockNumber) < 1000)
+        ) {
+          const transactionWithStatus = await this.getBlocksService.getTransaction(txPoint.txHash)
+          const ckbTransaction: CKBComponents.Transaction = transactionWithStatus.transaction
+          const transaction: Transaction = TypeConvert.toTransaction(ckbTransaction)
+          // tx timestamp / blockNumber / blockHash
+          const { blockHash } = transactionWithStatus.txStatus
+          if (blockHash) {
+            const blockHeader = await this.getBlocksService.getHeader(blockHash)
+            transaction.blockHash = blockHash
+            transaction.blockNumber = blockHeader.number
+            transaction.timestamp = blockHeader.timestamp
+          }
+          // broadcast address used
+          const txEntity = await TransactionPersistor.saveFetchTx(transaction)
+
+          let address: string | undefined
+          if (type === TxPointType.CreatedBy) {
+            address = LockUtils.lockScriptToAddress(transaction.outputs![+txPoint.index].lock)
+          } else if (type === TxPointType.ConsumedBy) {
+            const input = txEntity.inputs[+txPoint.index]
+            const output = await IndexerTransaction.updateInputLockHash(input.outPointTxHash!, input.outPointIndex!)
+            if (output) {
+              address = LockUtils.lockScriptToAddress(output.lock)
+            }
+          }
+          if (address) {
+            AddressesUsedSubject.getSubject().next([address])
+          }
+        }
+      }
+      page += 1
+    }
+  }
+
+  public stop = () => {
+    this.tipNumberListener.unsubscribe()
+    this.stopped = true
+  }
+
+  public waitForDrained = async (timeout: number = 5000) => {
```

Comment:
what does `waitForDrained` do?

Reply:
Ensure this queue stopped after call `stop`, then you can run a new queue.